docs/lang/articles/basic/type.md

```diff
@@ -1,217 +1,188 @@
 
-`ti.cast()` is used for normal type casting as in other programming languages:
+  ```python
+  @ti.kernel
+  def foo():
+      a = 3.14
+      b = ti.cast(a, ti.i32)  # 3
+      c = ti.cast(b, ti.f32)  # 3.0
+  ```
 
-```python {4-5}
-@ti.kernel
-def foo():
-    a = 3.14
-    b = ti.cast(a, ti.i32)  # 3
-    c = ti.cast(b, ti.f32)  # 3.0
-```
+- You can also use Python's builtin `int()` or `float()` to convert a value to the default integer type or to the default floating-point type:
 
-You can also use `int()` and `float()` to convert values to default integer and floating-point
-types:
+  ```python
+  @ti.kernel
+  def foo():
+      a = 3.14
+      b = int(a)    # 3
+      c = float(b)  # 3.0
+  ```
 
-```python {4-5}
-@ti.kernel
-def foo():
-    a = 3.14
-    b = int(a)    # 3
-    c = float(b)  # 3.0
-```
+### Implicit type casting
 
-#### Bit casting
+Implicit type casting occurs when you *accidentally* put or assign a value in a place where a different data type is expected.
 
-Use `ti.bit_cast()` to cast a value into another type **with its underlying bits preserved**:
+:::caution WARNING
+As a rule of thumb, implicit type casting is a major source of bugs. And Taichi does *not* recommend resorting to this mechanism.
 
-```python {4-5}
-@ti.kernel
-def foo():
-    a = 3.14
-    b = ti.bit_cast(a, ti.i32)  # 1078523331
-    c = ti.bit_cast(b, ti.f32)  # 3.14
-```
+:::
 
-Note that the new type must have the same precision bits as the old type (`i32`->`f64` is not
-allowed). Use this operation with caution.
+#### Implicit type casting in binary operations
 
-:::note
-`ti.bit_cast` is equivalent to `reinterpret_cast` in C++.
-:::
+Taichi follows the [implicit conversion rules](https://en.cppreference.com/w/c/language/conversion) for the C programming language and implicitly casts operands in a [binary operation](https://en.wikipedia.org/wiki/Binary_operation) into a *common type* if the operation involves different data types. Following are two most straightforward rules for determining the common type in a binary operation:
 
-### Implicit type casting
+- `i32 + f32 = f32` (`int` + `float` = `float`)
+- `i32 + i64 = i64` (low precision bits + high precision bits = high precision bits)
 
-When you accidentally use a value in a place where a different type is expected, implicit type
-casting is triggered for the following cases.
+#### Implicit type casting in assignments
 
-:::caution
-Relying on implicit type casting is bad practice and one major source of bugs.
-:::
+When you assign a value to a variable of a different data type, Taichi implicitly casts the value into that type. Further, if the type of the variable is *not* the common type, a warning of precision loss occurs.
 
-#### In binary operations
+- Example 1: Variable `a` is initialized with type `float` and immediately reassigned `1`. The reassignment implicitly casts `1` from `int` to `float` without warning because the data type of `a` is the common type `float`:
```

Comment:
reassigned with `1`?

Reply:
Active voice: Reassign a <value> to a <variable>; Passive voice. A <variable> is reassigned a <value>.